Add basic parallel.foreach to speed up processing documents in Roslyn Conventions

paulegradie opened the pull request:
# Description

In large codebases, there can be hundreds of thousands of unfiltered syntax nodes to iterate over when processing solution documents. This can take over a minute when applying a roslyn convention in a test.

Despite the extension of the DiagnosticAnalyzer class, it appears that the syntax tree filtering that typically happens when applying diagnostic analyzers does not seem to happen when applying these analyzers manually. The result is that the entire syntax tree is provided to the framework.

This PR implements a naive parallel.foreach to process the syntax tree more quickly.

## Alternatives

- If we can find a way to apply the node type filters via conventional - that should be the approach
- Barring that, if there is a preferred parallelization approach - I'd be happy to go with that

andrewabest commented:
This one feels like it needs a bit of thinking before we go ahead.

Roslyn analyzers implemented via `DiagnosticAnalyzer` allow us to register all sorts of analysis actions at various levels. 

Conventional takes a naive approach - hand it the solution, and it will review it to see if it complies to a given convention that has access to the entire syntax tree, and do this for each convention. It uses roslyn in a similar way to how it uses `Mono.Cecil` in the core implementation - as an object model that represents the codebase it can inspect.

Its understandable that at analyzer runtime this could be a bit sloooow.

It would be good to spike whether we could build the required analyzer context / tree once and reuse it across convention invocations, allowing for any sort of `DiagnosticAnalyzer` to be used. Ideally we want your IDE to use analyzer implementations that it can ensure are invoked during its compilation, and we then compose them into conventions that get applied to the solution.

andrewabest commented:
https://github.com/dotnet/roslyn-sdk/blob/main/samples/CSharp/Analyzers/Analyzers.Implementation/StatefulAnalyzers/CompilationStartedAnalyzerWithCompilationWideAnalysis.cs may be an option for us - roslyn analyzers that act a bit more like conventions in that they are applied to the whole codebase and run at the end.

empowerPaul commented:
Cool 👍 I'll look into this and come up with something.

paulegradie commented:
Hey @andrewabest, I've been looking into this and I've noticed a few things maybe you'll have additional insights on that I'm missing:

1. The two DiagnosticAnalyzerSpecifications provided (IfAndElse, and UsingStatements) are decorated with the `DiagnosticAnalyzer` attribute - but they don't inherit from the base DiagnosticAnalyzer class. These specs aren't going to be picked up as analyzers, so the attributes can be removed.

2. We currently generate the build workspace and manually iterate over each project, document, and syntax tree in each document. Some Solutions are of course going to have dozens of projects and thousands of documents, which could have extensive sytanx trees to traverse. Normally, an IDE is going to execute analyzers concurrently for us (when enabled - which should not be the default depending on how folks implement their analyzers), so we'll need to somehow enable concurrency in Conventional. I've currently implemented a slightly different approach that will let the runtime spin up threads when concurrency is enabled.
 
3. The current set of analyzer specs also hard code a syntax kind this isn't going to be suitable for all analyzers - so I've replaced those with an abstract method to return whatever is specific to the analyzer, which now has a public viritual. This wasn't causing a problem for the specs previously since the initialize method isn't invoked via the conventional framework - but now that we're filtering nodes to speed up processing, I've fixed that up.

Summary:
 - Since we're processing documents ourselves, we have to 
   - orchestrate concurrency
   - perform any node filtering
   - impement the analyzers in a way that allows both the conventional framework as we all compiler to execute the analyzers
   - decorate the analyzers (and not bother decorating the specs, which should simply invoke the analyzers manually)
  
I don't think this introduces any breaking changes - but it should enable much faster concurrent execution.

Regarding the SolutionWideAnalysis example - from what I understand this is still hooking into IDE/buildprocess tooling, which feels beyond the scope of what I'm trying to achieve here. Seems like we'd like to implement the logic to do thiat in conventional - when all we really need is to more quickly get through the many documents that the solution may hold.

andrewabest commented on src/Roslyn/Conventional.Roslyn/Conventions/SolutionDiagnosticAnalyzerConventionSpecification.cs:
```diff
@@ -10,26 +12,43 @@ public interface ISolutionDiagnosticAnalyzerConventionSpecification
-                .SelectMany(IsSatisfiedBy);
+            if (Analyzer.EnableConcurrentExecution())
+            {
+                var tasks = solution
```
I'd suggest we use `Parallel.ForEach` here to limit how wide we go with task creation.

andrewabest commented:
> Hey @andrewabest, I've been looking into this and I've noticed a few things maybe you'll have additional insights on that I'm missing:
> 
> 1. The two DiagnosticAnalyzerSpecifications provided (IfAndElse, and UsingStatements) are decorated with the `DiagnosticAnalyzer` attribute - but they don't inherit from the base DiagnosticAnalyzer class. These specs aren't going to be picked up as analyzers, so the attributes can be removed.
> 2. We currently generate the build workspace and manually iterate over each project, document, and syntax tree in each document. Some Solutions are of course going to have dozens of projects and thousands of documents, which could have extensive sytanx trees to traverse. Normally, an IDE is going to execute analyzers concurrently for us (when enabled - which should not be the default depending on how folks implement their analyzers), so we'll need to somehow enable concurrency in Conventional. I've currently implemented a slightly different approach that will let the runtime spin up threads when concurrency is enabled.
> 3. The current set of analyzer specs also hard code a syntax kind this isn't going to be suitable for all analyzers - so I've replaced those with an abstract method to return whatever is specific to the analyzer, which now has a public viritual. This wasn't causing a problem for the specs previously since the initialize method isn't invoked via the conventional framework - but now that we're filtering nodes to speed up processing, I've fixed that up.
> 
> Summary:
> 
> * Since we're processing documents ourselves, we have to
>   
>   * orchestrate concurrency
>   * perform any node filtering
>   * impement the analyzers in a way that allows both the conventional framework as we all compiler to execute the analyzers
>   * decorate the analyzers (and not bother decorating the specs, which should simply invoke the analyzers manually)
> 
> I don't think this introduces any breaking changes - but it should enable much faster concurrent execution.
> 
> Regarding the SolutionWideAnalysis example - from what I understand this is still hooking into IDE/buildprocess tooling, which feels beyond the scope of what I'm trying to achieve here. Seems like we'd like to implement the logic to do thiat in conventional - when all we really need is to more quickly get through the many documents that the solution may hold.

Thanks for taking the time to dig into this!

I think we could actually remove `EnableConcurrentExecution` to start with and just execute the analyzers in parallel every time during tests. It would be hard to reason about the impact of parallel execution if each analyzer is opting in or out of it - it is simpler for us to just make this the default way of executing the conventions.

My only suggestion aside from simplifying the approach is to use `Parallel.ForEach` so that we get some respect for the capacity of the system running the tests in the amount we actually parallelize :)

paulegradie commented:
I appreciate the feedback on this 👍 

I've made the suggested changes on:
 - concurrency in the analyzer
 - parallel.foreach

Couple of quick notes to tie things off:
 - the compiler for some reason things the spec is an analyzer (maybe because of the project name or references), so I've supressed an `RS1035`
 - Happy to remove the option to disable concurrent execution, but something to consider is if folks ever want to disable concurrency in a debug scenario. Probably moot since they'll have already validated their analyzer logic in conventional.

Once this merges, Conventional will be 🏎️🔥!

As usual, happy to make any additional changes you prefer on this 👋

andrewabest reviewed (approved):
Looks good Paul. This will be a major version bump due to the API change, so look for the 12.x version to test it out after it has merged.
